resources/type-system/upper-lower-bounds.md

```diff
@@ -0,0 +1,390 @@
+# Dart 2.0 Upper and Lower bounds
+
+leafp@google.com
+
+This documents the currently implemented upper and lower bound computation,
+modified to account for explicit nullability and the accompanying type system
+changes (including the legacy types).  In the interest of backwards
+compatibility, it does not try to fix the various issues with the existing
+algorithm.  
+
+## Types
+
+The syntactic set of types used in this draft are a slight simplification of
+full Dart types, as described in the subtyping
+document
+[here](https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md)
+
+## Helper predicates
+
+The **TOP** predicate is true for any type which is in the equivalence class of
+top types.
+
+- **TOP**(`T?`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`T*`) is true iff **TOP**(`T`) or **OBJECT**(`T`)
+- **TOP**(`dynamic`) is true
+- **TOP**(`void`) is true
+- **TOP**(`FutureOr<T>`) is **TOP**(T)
+- **TOP**(T) is false otherwise
+
+The **OBJECT** predicate is true for any type which is in the equivalence class
+of `Object`.
+
+
+- **OBJECT**(`Object`) is true
+- **OBJECT**(`dynamic`) is true
+- **OBJECT**(`void`) is true
+- **OBJECT**(`FutureOr<T>`) is **OBJECT**(T)
+- **OBJECT**(T) is false otherwise
+
+The **BOTTOM** predicate is true for things in the equivalence class of `Never`.
+
+- **BOTTOM**(`Never`) is true
+- **BOTTOM**(`X&T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`X extends T`) is true iff **BOTTOM**(`T`)
+- **BOTTOM**(`T`) is false otherwise
+
+The **NULL** predicate is true for things in the equivalence class of `Null`
+
+- **NULL**(`Null`) is true
+- **NULL**(`T?`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T*`) is true iff **NULL**(`T`) or **BOTTOM**(`T`)
+- **NULL**(`T`) is false otherwise
+
+The **MORETOP** predicate defines a total order on top and `Object` types.
+
+- **MORETOP**(`void`, `T`) = true
+- **MORETOP**(`T`, `void`) = false
+- **MORETOP**(`dynamic`, `T`) = true
+- **MORETOP**(`T`, `dynamic`) = false
+- **MORETOP**(`Object`, `T`) = true
```

Comment:
I do not understand why this would be `Object` rather than `Object?`. With `Object` we get things like `MORETOP(Object, Object?)`, and that definitely seems wrong.

Reply:
Yeah, this is all a bit annoying.  I think it works as written, because:
  - You only initially call this when you either have two types satisfying **TOP** or two types satisfying **OBJECT**
  - Within that constraint, the only other thing that matters is that we impose a total order, which we do.

So you are correct, that we have **MORETOP**(`Object`, `Object?`), but you never ask that question directly.  Instead, you must have asked something like **MORETOP**(`Object?`, `Object??`), which recursively reduces to the above question.  So we get the answer we want, just in a weird way.

I'll think about whether I can split this out.  It feels a bit ugly though - I need a total order on all TOP types, and that means that I recursively I need to be able order an OBJECT type relative to a TOP type.

Reply:
Couldn't **MORETOP** use `Object?` in the two cases I mentioned, and then we'd have a `**MOREOBJECT**` which is simply

**MOREOBJECT(FutureOr^{k+1}<Object>, FutureOr^k<Object>) for all k >= 0.

and then we wouldn't need to look outside the definition of **MORETOP** to see that **MORETOP**(Object, Object?)` won't happen.

Reply:
I think what you are asking is: "why not have a MORETOP which only applies to the top types, and a MOREOBJECT which only applies to the Object types".  I think you can do this, but then you need to do a bunch of work to make sure that MORETOP doesn't get recursively called on Object types.  It's not enough to just pattern match on `Object?`, because you can have `FutureOr<Object>?` as an argument, and you need to be sure not to recursively call MORETOP on that, because it's not a top type.  I think to make this work, you do something like this:

**MORETOP**(`T?`, `S?`) where **TOP**(`T`) and **TOP**(`S`)  = **MORETOP**(`T`, `S`)
**MORETOP**(`T?`, `S?`) where **OBJECT**(`T`) and **OBJECT**(`S`) = **MOREOBJECT**(`T`, `S`)
**MORETOP**(`T?`, `S?`) where **OBJECT**(`T`) and **TOP**(`S`) = true
**MORETOP**(`T?`, `S?`) where **TOP**(`T`) and **OBJECT**(`S`) = false

and similarly for `T*` and `S*`.

I think that works fine as well.  It's a little more verbose.  And it's kind of inefficient to be checking for TOP and OBJECT on each recursive call, but in practice I doubt it really matters.  I could be convinced to specify it that way though.  Thoughts?